File: GCDLib/Core/Visualization/DoDHistogramViewerClass.vb
Imports System.Windows.Forms.DataVisualization
Imports UnitsNet

Namespace Core.Visualization

    Public Class DoDHistogramViewerClass

        ' Names for data series
        Private Const EROSION As String = "Erosion"
        Private Const DEPOSITION As String = "Deposition"
        Private Const RAW As String = "Raw"

        Private m_Chart As Charting.Chart

        ' Raw histogram data
        Private m_Raw As Dictionary(Of Double, ElevationChangeDataPoint)
        Private m_Thresholded As Dictionary(Of Double, ElevationChangeDataPoint)
        Private m_LinearDataUnits As Units.LengthUnit

        Private m_LinearDisplayUnits As Units.LengthUnit
        Private m_AreaDisplayUnits As Units.AreaUnit
        Private m_VolumeDisplayUnits As Units.VolumeUnit

        Private ReadOnly Property AreaDataUnits
            Get
                Return GCDConsoleLib.Utility.Conversion.LengthUnit2AreaUnit(m_LinearDataUnits)
            End Get
        End Property

        Private ReadOnly Property VolumeDataUnits
            Get
                Return GCDConsoleLib.Utility.Conversion.LengthUnit2VolumeUnit(m_LinearDataUnits)
            End Get
        End Property

        ''' <summary>
        ''' Call this constructor from non-UI code that simply wants to generate histogram plot image files
        ''' </summary>
        Public Sub New(sRawHistogram As String, sThreshHistogram As String, linearDataUnits As UnitsNet.Units.LengthUnit)
            m_Chart = New Charting.Chart
            Init(sRawHistogram, sThreshHistogram, linearDataUnits)
        End Sub

        ''' <summary>
        ''' Constructor for UI code to pass in a chart on a user interface form
        ''' </summary>
        ''' <param name="cht"></param>
        Public Sub New(ByRef cht As Charting.Chart, sRawHistogram As String, sThreshHistogram As String, linearDataUnits As UnitsNet.Units.LengthUnit)
            m_Chart = cht
            Init(sRawHistogram, sThreshHistogram, linearDataUnits)
        End Sub

        Private Sub Init(sRawHistogram As String, sThreshHistogram As String, linearDataUnits As UnitsNet.Units.LengthUnit)

            ' Load the data for both the raw and thresholded histograms
            LoadHistogram(sRawHistogram, m_Raw)
            LoadHistogram(sThreshHistogram, m_Thresholded)
            m_LinearDataUnits = linearDataUnits

            m_LinearDisplayUnits = m_LinearDataUnits
            m_AreaDisplayUnits = AreaDataUnits
            m_VolumeDisplayUnits = VolumeDataUnits

            ' Proceed and do the one-time chart preparation
            m_Chart.ChartAreas.Clear()
            m_Chart.ChartAreas.Add(New Charting.ChartArea())

            m_Chart.Series.Clear()
            m_Chart.Palette = Nothing
            m_Chart.Legends.Clear()

            Dim seriesDefs = New Dictionary(Of String, System.Drawing.Color) From {{EROSION, My.Settings.Erosion}, {DEPOSITION, My.Settings.Depsoition}, {RAW, Drawing.Color.LightGray}}
            For Each aDef As KeyValuePair(Of String, Drawing.Color) In seriesDefs
                Dim series As Charting.Series = m_Chart.Series.Add(aDef.Key)
                series.ChartType = Charting.SeriesChartType.StackedColumn
                series.Color = aDef.Value
                series.ChartArea = m_Chart.ChartAreas.First().Name
            Next

            With m_Chart.ChartAreas(0).AxisX
                .MajorGrid.LineColor = Drawing.Color.LightSlateGray
                .MinorTickMark.Enabled = True
            End With

            With m_Chart.ChartAreas(0).AxisY
                .MajorGrid.LineColor = Drawing.Color.LightSlateGray
                .MinorTickMark.Enabled = True
            End With

            RefreshDisplay(True, linearDataUnits, GCDConsoleLib.Utility.Conversion.LengthUnit2AreaUnit(linearDataUnits), GCDConsoleLib.Utility.Conversion.LengthUnit2VolumeUnit(linearDataUnits))

        End Sub

        Public Sub SetChartType(bArea As Boolean)
            RefreshDisplay(bArea, m_LinearDisplayUnits, m_AreaDisplayUnits, m_VolumeDisplayUnits)
        End Sub

        Public Sub RefreshDisplay(ByVal bArea As Boolean, ByVal linearDisplayUnits As UnitsNet.Units.LengthUnit, areaDisplayUnits As UnitsNet.Units.AreaUnit, volumeDisplayUnits As UnitsNet.Units.VolumeUnit)

            ' Store the display units so that the user can switch between area and volume easily
            m_LinearDisplayUnits = linearDisplayUnits
            m_AreaDisplayUnits = areaDisplayUnits
            m_VolumeDisplayUnits = volumeDisplayUnits

            Dim histoData As List(Of HistogramDisplayDataPoint) = GetDisplayValues(bArea, linearDisplayUnits, areaDisplayUnits, volumeDisplayUnits)

            m_Chart.Series.FindByName(EROSION).Points.DataBindXY(histoData, "elevation", histoData, "erosion")
            m_Chart.Series.FindByName(DEPOSITION).Points.DataBindXY(histoData, "elevation", histoData, "deposition")
            m_Chart.Series.FindByName(RAW).Points.DataBindXY(histoData, "elevation", histoData, "raw")

            With m_Chart.ChartAreas(0)
                .AxisX.Title = String.Format("Elevation Change ({0})", linearDisplayUnits)

                If bArea Then
                    .AxisY.Title = String.Format("Area ({0}²)", UnitsNet.Area.GetAbbreviation(areaDisplayUnits))
                Else
                    .AxisY.Title = String.Format("Volume ({0}³)", UnitsNet.Volume.GetAbbreviation(volumeDisplayUnits))
                End If
            End With

        End Sub

        Private Function GetDisplayValues(bArea As Boolean, linearDisplayUnits As Units.LengthUnit, areaDisplayUnits As Units.AreaUnit, volDisplayUnits As Units.VolumeUnit) As List(Of HistogramDisplayDataPoint)

            ' Note that the key to this dictionary is the histogram elevation values in their ORIGINAL units
            ' while the elevation properties of the HistogramDisplayDataPoint should be in the display units
            Dim lDisplayData As New Dictionary(Of Double, HistogramDisplayDataPoint)

            ' There must always be a thresholded histogram that is displayed red/blue
            For Each dataPoint As ElevationChangeDataPoint In m_Thresholded.Values
                lDisplayData(dataPoint.Elevation) = New HistogramDisplayDataPoint(dataPoint.GetElevation(m_LinearDataUnits, linearDisplayUnits))

                If bArea Then
                    lDisplayData(dataPoint.Elevation).Erosion = dataPoint.AreaErosion(AreaDataUnits, areaDisplayUnits)
                    lDisplayData(dataPoint.Elevation).Deposition = dataPoint.AreaDeposition(AreaDataUnits, areaDisplayUnits)
                Else
                    lDisplayData(dataPoint.Elevation).Erosion = dataPoint.VolumeErosion(VolumeDataUnits, volDisplayUnits)
                    lDisplayData(dataPoint.Elevation).Deposition = dataPoint.VolumeDeposition(VolumeDataUnits, volDisplayUnits)
                End If
            Next

            ' If there's a raw histogram then load the values. These will be displayed as a grey column
            If Not m_Raw Is Nothing Then
                For Each item As KeyValuePair(Of Double, ElevationChangeDataPoint) In m_Raw
                    If Not lDisplayData.ContainsKey(item.Key) Then
                        lDisplayData(item.Key) = New HistogramDisplayDataPoint(Length.From(item.Value.Elevation, m_LinearDataUnits).As(linearDisplayUnits))
                    End If

                    If item.Key < 0 Then
                        If bArea Then
                            lDisplayData(item.Key).Raw = Math.Max(0, item.Value.AreaChange(AreaDataUnits, areaDisplayUnits) - lDisplayData(item.Key).Erosion)
                        Else
                            lDisplayData(item.Key).Raw = Math.Max(0, item.Value.VolumeChange(volDisplayUnits, volDisplayUnits) - lDisplayData(item.Key).Erosion)
                        End If
                    Else
                        If bArea Then
                            lDisplayData(item.Key).Raw = Math.Max(0, item.Value.AreaChange(AreaDataUnits, areaDisplayUnits) - lDisplayData(item.Key).Deposition)
                        Else
                            lDisplayData(item.Key).Raw = Math.Max(0, item.Value.VolumeChange(volDisplayUnits, volDisplayUnits) - lDisplayData(item.Key).Deposition)
                        End If
                    End If
                Next
            End If

            Return lDisplayData.Values.ToList()

        End Function

        Public Sub ExportCharts(ByVal AreaGraphPath As String, ByVal VolumeGraphPath As String, ByVal ChartWidth As Integer, ByVal ChartHeight As Integer)

            If Not IO.Directory.Exists(IO.Path.GetDirectoryName(AreaGraphPath)) Then
                Dim ex As New Exception("The output folder for the GCD area graph does not exist.")
                ex.Data("Area Graph Path") = AreaGraphPath
            End If

            If Not IO.Directory.Exists(IO.Path.GetDirectoryName(VolumeGraphPath)) Then
                Dim ex As New Exception("The output folder for the GCD volume graph does not exist.")
                ex.Data("volume Graph Path") = VolumeGraphPath
            End If

            RefreshDisplay(True, m_LinearDataUnits, AreaDataUnits, VolumeDataUnits)
            m_Chart.SaveImage(AreaGraphPath, Charting.ChartImageFormat.Png)

            RefreshDisplay(False, m_LinearDataUnits, AreaDataUnits, VolumeDataUnits)
            m_Chart.SaveImage(VolumeGraphPath, Charting.ChartImageFormat.Png)

        End Sub

        ''' <summary>
        ''' Loads a histogram from a CSV
        ''' </summary>
        ''' <param name="csvFilePath">Input path to CSV file</param>
        ''' <remarks></remarks>
        Private Sub LoadHistogram(ByVal csvFilePath As String, ByRef values As Dictionary(Of Double, ElevationChangeDataPoint))

            Try
                If Not IO.File.Exists(csvFilePath) Then
                    Throw New Exception("Elevation change histogram CSV file does not exist.")
                End If

                values = New Dictionary(Of Double, ElevationChangeDataPoint)

                Using readFile As New IO.StreamReader(csvFilePath)

                    ' skip first headers line
                    Dim line As String = readFile.ReadLine()
                    Dim csvdata As String()
                    While True
                        line = readFile.ReadLine()
                        If line Is Nothing Then
                            Exit While
                        Else
                            csvdata = Split(line, ",")
                            Dim fElevation As Double = Double.Parse(csvdata(0))
                            values(fElevation) = New ElevationChangeDataPoint(fElevation, Double.Parse(csvdata(2)), Double.Parse(csvdata(3)))
                        End If
                    End While
                End Using

            Catch ex As IO.IOException
                Dim ex2 As New Exception("Could not access elevation change histogram file because it is being used by another program.", ex)
                ex2.Data("File Path") = csvFilePath
                Throw ex2
            Catch ex As Exception

            End Try
        End Sub

    End Class

End Namespace

File: GCDCore/Visualization/DoDHistogramViewerClass.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms.DataVisualization.Charting;
using System.IO;
using System.Drawing;
using GCDConsoleLib;

namespace GCDCore.Visualization
{
    public class DoDHistogramViewerClass
    {
        // Names for data series
        private const string EROSION = "Erosion";
        private const string DEPOSITION = "Deposition";
        private const string RAW = "Raw";

        private Chart m_Chart;

        // Raw histogram data
        Histogram _rawHist;
        Histogram _thrHist;

        private Dictionary<decimal, HistogramDisplayData> histoData;

        private readonly GCDConsoleLib.GCD.UnitGroup DataUnits;
        private GCDConsoleLib.GCD.UnitGroup DisplayUnits { get; set; }


        /// <summary>
        /// NOTE: The decimals in here must already be in their display unit
        /// </summary>
        public class HistogramDisplayData
        {
            public decimal Threshold { get; set; }
            public decimal Raw { get; set; }
            public decimal Elevation { get; private set; }

            public decimal Deposition { get { return Elevation > 0 ? Threshold : 0; } }
            public decimal Erosion { get { return Elevation < 0 ? Threshold : 0; } }

            public HistogramDisplayData(decimal elev)
            {
                Elevation = elev;
                Threshold = 0;
            }
        }

        /// <summary>
        /// Call this constructor from non-UI code that simply wants to generate histogram plot image files
        /// </summary>
        public DoDHistogramViewerClass(Histogram rawHisto, Histogram thrHisto, GCDConsoleLib.GCD.UnitGroup dataUnits, Chart chtControl = null)
        {
            DataUnits = dataUnits;
            DisplayUnits = dataUnits;

            // Proceed and do the one-time chart preparation
            if (chtControl == null)
                m_Chart = new Chart();
            else
                m_Chart = chtControl;

            m_Chart.ChartAreas.Clear();
            m_Chart.ChartAreas.Add(new ChartArea());

            m_Chart.Series.Clear();
            m_Chart.Palette = ChartColorPalette.None;
            m_Chart.Legends.Clear();

            Dictionary<string, Color> seriesDefs = new Dictionary<string, Color> {
                {  EROSION, Properties.Settings.Default.Erosion },
                {  DEPOSITION, Properties.Settings.Default.Deposition  },
                { RAW, Color.LightGray }
            };

            foreach (KeyValuePair<string, Color> aDef in seriesDefs)
            {
                Series series = m_Chart.Series.Add(aDef.Key);
                series.ChartType = SeriesChartType.StackedColumn;
                series.Color = aDef.Value;
                series.ChartArea = m_Chart.ChartAreas.First().Name;
            }

            var _with1 = m_Chart.ChartAreas[0].AxisX;
            _with1.MajorGrid.LineColor = Color.LightSlateGray;
            _with1.MinorTickMark.Enabled = true;

            var _with2 = m_Chart.ChartAreas[0].AxisY;
            _with2.MajorGrid.LineColor = Color.LightSlateGray;
            _with2.MinorTickMark.Enabled = true;

            _rawHist = rawHisto;
            _thrHist = thrHisto;

            UpdateDisplay(true);
        }

        public void SetChartType(bool bSetArea)
        {
            UpdateDisplay(bSetArea);
        }

        public void UpdateDisplay(bool bArea, GCDConsoleLib.GCD.UnitGroup displayUnits = null)
        {
            // Store the display units so that the user can switch between area and volume easily
            if (displayUnits != null)
                DisplayUnits = displayUnits;

            // Go recalc our values
            GetDisplayValues(bArea);

            m_Chart.Series.FindByName(EROSION).Points.DataBindXY(histoData.Values, "Elevation", histoData.Values, "Erosion");
            m_Chart.Series.FindByName(DEPOSITION).Points.DataBindXY(histoData.Values, "Elevation", histoData.Values, "Deposition");
            m_Chart.Series.FindByName(RAW).Points.DataBindXY(histoData.Values, "Elevation", histoData.Values, "Raw");

            var _with3 = m_Chart.ChartAreas[0];
            _with3.AxisX.Title = string.Format("Elevation Change ({0})", UnitsNet.Length.GetAbbreviation(DisplayUnits.VertUnit));

            if (bArea)
                _with3.AxisY.Title = string.Format("Area ({0})", UnitsNet.Area.GetAbbreviation(DisplayUnits.ArUnit));
            else
                _with3.AxisY.Title = string.Format("Volume ({0})", UnitsNet.Volume.GetAbbreviation(DisplayUnits.VolUnit));
        }

        private void GetDisplayValues(bool bArea)
        {
            // Note that the key to this dictionary is the histogram elevation values in their ORIGINAL units
            // while the elevation properties of the HistogramDisplayDataPoint should be in the display units
            histoData = new Dictionary<decimal, HistogramDisplayData>();

            for (int bid = 0; bid < _thrHist.Count; bid++)
            {
                // Make a dictionary entry if we don't already have one
                decimal binleft = (decimal)_thrHist.BinLower(bid, DataUnits).As(DisplayUnits.VertUnit);
                if (!histoData.ContainsKey(binleft)) histoData[binleft] = new HistogramDisplayData(binleft);

                if (bArea)
                    histoData[binleft].Threshold = (decimal)(_thrHist.BinArea(bid, Project.ProjectManager.Project.CellArea).As(DisplayUnits.ArUnit));
                else
                    histoData[binleft].Threshold = (decimal)_thrHist.BinVolume(bid, Project.ProjectManager.Project.CellArea, DataUnits).As(DisplayUnits.VolUnit);
            }

            if ((_rawHist != null))
            {
                for (int bid = 0; bid < _rawHist.Count; bid++)
                {
                    // Make a dictionary entry if we don't already have one
                    decimal binleft = (decimal)_rawHist.BinLower(bid, DataUnits).As(DisplayUnits.VertUnit);
                    if (!histoData.ContainsKey(binleft)) histoData[binleft] = new HistogramDisplayData(binleft);

                    if (bArea)
                        histoData[binleft].Raw = (decimal)(_rawHist.BinArea(bid, Project.ProjectManager.Project.CellArea).As(DisplayUnits.ArUnit)) - histoData[binleft].Threshold;
                    else
                        histoData[binleft].Raw = (decimal)_rawHist.BinVolume(bid, Project.ProjectManager.Project.CellArea, DataUnits).As(DisplayUnits.VolUnit) - histoData[binleft].Threshold;
                }

            }

        }

        public void ExportCharts(string AreaGraphPath, string VolumeGraphPath, int ChartWidth, int ChartHeight)
        {
            if (!Directory.Exists(Path.GetDirectoryName(AreaGraphPath)))
            {
                Exception ex = new Exception("The output folder for the GCD area graph does not exist.");
                ex.Data["Area Graph Path"] = AreaGraphPath;
            }

            if (!Directory.Exists(Path.GetDirectoryName(VolumeGraphPath)))
            {
                Exception ex = new Exception("The output folder for the GCD volume graph does not exist.");
                ex.Data["volume Graph Path"] = VolumeGraphPath;
            }

            UpdateDisplay(true, DataUnits);
            m_Chart.SaveImage(AreaGraphPath, ChartImageFormat.Png);

            UpdateDisplay(false, DataUnits);
            m_Chart.SaveImage(VolumeGraphPath, ChartImageFormat.Png);
        }
    }
}